src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -0,0 +1,139 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using NuGet.Common;
+using NuGet.LibraryModel;
+using NuGet.Packaging.Core;
+using NuGet.Shared;
+
+namespace NuGet.ProjectModel
+{
+    public static class PackagesLockFileUtilities
+    {
+        public static bool IsNuGetLockFileSupported(PackageSpec project)
+        {
+            var restorePackagesWithLockFile = project.RestoreMetadata?.RestoreLockProperties.RestorePackagesWithLockFile;
+            return MSBuildStringUtility.IsTrue(restorePackagesWithLockFile) || File.Exists(GetNuGetLockFilePath(project));
+        }
+
+        public static string GetNuGetLockFilePath(PackageSpec project)
+        {
+            if (project.RestoreMetadata == null || project.BaseDirectory == null)
+            {
+                // RestoreMetadata or project BaseDirectory is not set which means it's probably called through test.
+                return null;
+            }
+
+            var path = project.RestoreMetadata.RestoreLockProperties.NuGetLockFilePath;
+
+            if (string.IsNullOrEmpty(path))
```

Comment:
When user doesn't set this `NuGetLockFilePath` property explicitly and instead wants to use the default file name n path then it will be null. This is not for tests. You can find more details in specs.

Reply:
If it's restore metadata, you should generate it at the beginning similar to what we do with the assets and cache files.

Reply:
again not able to understand your concerns here as well... may be we speak different language. its to know the lock file path when we start to evaluate it now i dont understand what you mean by at the beginning...?

Again lets discuss offline